help_to_heat/frontdoor/routing/forwards_routing.py

```diff
@@ -0,0 +1,500 @@
+    country_field_northern_ireland,
+    country_field_scotland,
+    country_field_wales,
+    country_page,
+    duplicate_uprn_field,
+    epc_accept_suggested_epc_field,
+    epc_found_field,
+    epc_ineligible_page,
+    epc_page,
+    epc_rating_is_eligible_field,
+    epc_select_choice_field,
+    epc_select_choice_field_enter_manually,
+    epc_select_choice_field_epc_api_fail,
+    epc_select_choice_field_select_epc,
+    epc_select_page,
+    field_no,
+    field_yes,
+    household_income_field,
+    household_income_page,
+    loft_access_field,
+    loft_access_page,
+    loft_field,
+    loft_field_no,
+    loft_field_yes,
+    loft_insulation_field,
+    loft_insulation_page,
+    loft_page,
+    northern_ireland_ineligible_page,
+    number_of_bedrooms_field,
+    number_of_bedrooms_page,
+    own_property_field,
+    own_property_field_social_housing,
+    own_property_fields_non_social_housing,
+    own_property_page,
+    park_home_field,
+    park_home_ineligible_page,
+    park_home_main_residence_field,
+    park_home_main_residence_page,
+    park_home_page,
+    property_ineligible_page,
+    property_subtype_field,
+    property_subtype_page,
+    property_type_field,
+    property_type_page,
+    referral_already_submitted_page,
+    schemes_page,
+    success_page,
+    summary_page,
+    supplier_field,
+    supplier_field_british_gas,
+    supplier_field_bulb,
+    supplier_field_e,
+    supplier_field_edf,
+    supplier_field_eon_next,
+    supplier_field_foxglove,
+    supplier_field_octopus,
+    supplier_field_ovo,
+    supplier_field_scottish_power,
+    supplier_field_shell,
+    supplier_field_utilita,
+    supplier_field_utility_warehouse,
+    supplier_page,
+    unknown_page,
+    utility_warehouse_warning_page,
+    utility_warehouse_warning_page_field,
+    wall_insulation_field,
+    wall_insulation_page,
+    wall_type_field,
+    wall_type_page,
+)
+from help_to_heat.frontdoor.eligibility import calculate_eligibility
+
+_unknown_response = unknown_page, False
+
+
+def _requires_answer(answer_key):
+    def wrapper(func):
+        def next_page_function(answers):
+            required_answer = answers.get(answer_key)
+            if required_answer is None:
+                return _unknown_response
+
+            return func(answers)
+
+        return next_page_function
+
+    return wrapper
+
+
+def get_next_page(current_page, answers):
+    """
+    Generates the next page ID in the flow, given the users' current page and answers.
+    Parameters
+    ----------
+    current_page
+        Page ID the user is currently on.
+    answers
+        All answers given by the user so far. Should include answers submitted on the current page
+
+    Returns
+    -------
+    (str | None, bool)
+        New page ID, or `unknown_page` (None) if the user hasn't given the required information on this page to
+        determine the next page.
+
+        Whether this next page should be redirected to immediately
```

Comment:
thinking about this more there is a more pressing use case. when we generate the previous page we need to know if any redirects happened, for instance

---> click
==> redirect

so if the route is `address` --->` epc-select` ==> `address select`, the back link on `address` should not link to `epc-select` (since that would immediately redirect back and cause a loop) but it should know to go back further to `address`. having the redirect state be returned means the back routing can know to make this jump

in a more broad sense having this redirect param means the routing can know to ignore the redirected page from the route (the user never 'goes' there)

Reply:
Sure, understood. I feel like this is probably the least-effort way to do it at the moment, but feels like a bit of a strange pattern so should interrogate:
> this API request reasonably has to happen on GET so we can't do any routing on POST to the preview page

I think we talked about this before, but I can't remember the conclusion. Why is this? On the face of it it feels more natural to make API requests based on input data and decide routing on POST.

Reply:
potentially, though it'd need a bit of a rework into the site's logic. currently the flow is
input address -> save address to session -> load epc-select page -> make API request based on address in session -> redirect if needed

we'd need to rework it to be
input address -> make API request based on address -> save returned EPCs to session -> load epc-select page -> display them

Reply:
have added as PC-1255